lib/graphql/batch/execution_strategy.rb

```diff
@@ -1,57 +1,62 @@
 module GraphQL::Batch
   class ExecutionStrategy < GraphQL::Query::SerialExecution
-    attr_reader :batched_queries
-
-    def initialize
-      @batched_queries = Hash.new{ |hash, key| hash[key] = [] }
-    end
-
     class OperationResolution < GraphQL::Query::SerialExecution::OperationResolution
       def result
-        result = super
-        until execution_strategy.batched_queries.empty?
-          queries = execution_strategy.batched_queries.shift.last
-          queries.first.class.execute(queries)
-        end
-        result
+        GraphQL::Batch.execute(super)
       end
     end
 
     class SelectionResolution < GraphQL::Query::SerialExecution::SelectionResolution
       def result
-        result_hash = super
-        result_hash.each do |key, value|
-          if value.is_a?(FieldResolution)
-            value.result_hash = result_hash
+        wrap_queries(super)
+      end
+
+      private
+
+      def wrap_queries(obj)
+        queries = nil
+        case obj
+        when Array
+          array = obj
+          array.each_with_index do |value, i|
+            value = wrap_queries(value)
+            if value.is_a?(QueryContainer)
+              array[i] = nil
+              queries ||= []
+              queries << value.then do |result|
+                array[i] = result
+              end
+            end
+          end
+        when Hash
+          hash = obj
+          hash.each do |key, value|
+            value = wrap_queries(value)
+            if value.is_a?(QueryContainer)
+              hash[key] = nil
+              queries ||= []
```

Comment:
I know. I was just trying to avoid extra object allocations, since this gets called on every value.